### PR TITLE
New TbotsRobotMsg protos

### LINKED ISSUE #1272: Create TbotsRobotMsg Proto
### Description of the task
Currently, robot statuses such as break beam, capacitor voltage, etc., are sent to the robot in custom encoded packets over radio. The new TbotsRobotMsg Proto will use the Protobuf encoding and standardize this packet for the old and new robots. The new proto design will need to consider the needs of the new robot including new sensors or new sensor data we want to collect and should be done with sufficient consultation.

### Acceptance criteria

- [ ] Develop structure for Tbots_RobotMsg Proto
- [ ] Consult with relevant stakeholders
- [ ] Finalize the proto

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Prototyped a new TbotsRobotMsg proto that will be returned by the new robots and will be part of the interface between Backend and SensorFusion

### Testing Done
Builds

### Resolved Issues
Resolves #1272

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/proto/robot_component_status.proto [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    // UINT64_MAX indicates chipper was never fired
+    optional uint64 time_since_chipper_fired = 2;
+
+    // Indicates the time elapsed since kicker was last fired in ms
+    // UINT64_MAX indicates kicker was never fired
+    optional uint64 time_since_kicker_fired = 3;
+
+    /* 2014 Robot Fields */
+    // Indicates when autokick has been fired.
+    optional bool autokick_fired = 8;
```
Comment (jonathanlew) [inline src/shared/proto/robot_component_status.proto]: thoughts on whether we want `time_since_chipper_fired` and  `time_since_kicker_fired` instead of `autokick_fired` for the new robots?

File: src/shared/proto/robot_component_status.proto [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+/* Firmware build information  */
+message FirmwareStatus
+{
+    // The microcontroller firmware build ID.
+    required uint32 fw_build_id = 1;
+
+    // Whether or not the build ID information is valid.
+    required bool build_ids_valid = 3;
```
Comment (jonathanlew) [inline src/shared/proto/robot_component_status.proto]: @garethellis0, @MathewMacDougall brought up whether we need this in addition to fw_build_id

File: src/shared/proto/tbots_robot_msg.proto [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    required uint32 robot_id                           = 1;
+    repeated ErrorCode error_code                      = 2;
+    required BreakBeamStatus break_beam_status         = 3;
+    required ChipperKickerStatus chipper_kicker_status = 4;
+    required MotorEncoderData motor_encoder_data       = 5;
+    required FirmwareStatus firmware_status            = 6;
+    required DribblerStatus dribbler_status            = 7;
+    required NetworkStatus network_status              = 8;
+    required PowerStatus power_status                  = 9;
+    required TemperatureStatus temperature_status      = 10;
```
Comment (jonathanlew) [inline src/shared/proto/tbots_robot_msg.proto]: Thoughts on whether `required` fields should be `optional` instead? (https://github.com/UBC-Thunderbots/Software/pull/1274#pullrequestreview-396471163)

**Comment by jonathanlew** [comment]
Reopening this proto PR because @garethellis0 couldn't comment on the other one. I removed the fields that were deemed unnecessary and migrated the outstanding discussion from #1274, but if I missed something, please bring it up again

**Comment by akhilveeraghanta** [comment]
Overall I really like the new structure, nice work fire!

We have a lot of required fields, this may or may not be an issue, but since we are following googles style guide, just wanted to highlight this comment they had on required vs optional

"You should be very careful about marking fields as required. If at some point you wish to stop writing or sending a required field, it will be problematic to change the field to an optional field – old readers will consider messages without this field to be incomplete and may reject or drop them unintentionally. You should consider writing application-specific custom validation routines for your buffers instead. Some engineers at Google have come to the conclusion that using required does more harm than good; they prefer to use only optional and repeated. However, this view is not universal."

File: src/shared/proto/tbots_robot_msg.proto [hunk 4]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto2";
+
+import "shared/proto/robot_component_status.proto";
+
+message TbotsRobotMsg
```
Comment (akhilveeraghanta) [inline src/shared/proto/tbots_robot_msg.proto:5]: Should we add a test that builds this message with all the fields required for the old robots and verifies that it doesn't go over 100 bytes like Gareth suggested? I feel like we are getting pretty close if not already over the limit

**Review by akhilveeraghanta — COMMENTED** [review]
Brought over 2 comments from the other PR

File: src/shared/proto/tbots_robot_msg.proto (hunk 4 above)
Comment (jonathanlew) [reply src/shared/proto/tbots_robot_msg.proto:5]: No, the current plan is to keep the current radio annunciator and convert those messages into TbotsRobotMsg in the RadioBackend

Comment (akhilveeraghanta) [reply src/shared/proto/tbots_robot_msg.proto:5]: Ok that’s fair

File: src/shared/proto/tbots_robot_msg.proto (hunk 3 above)
Comment (akhilveeraghanta) [reply src/shared/proto/tbots_robot_msg.proto]: Id vouch for optional everywhere for consistency sake (as well as the reasons from (https://github.com/UBC-Thunderbots/Software/pull/1274#pullrequestreview-396471163))

`NetworkStatus` for example has fields that are entirely optional, but the `network_status` msg itself is required?

File: src/shared/proto/robot_component_status.proto (hunk 2 above)
Comment (akhilveeraghanta) [reply src/shared/proto/robot_component_status.proto]: This should probably be 
```suggestion
    // Whether or not the build ID information is valid.
    required bool build_ids_valid = 2;
```

File: src/shared/proto/robot_component_status.proto [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    required float break_beam_reading = 2;
+}
+
+/* Data about the status of the chipper and kicker */
+message ChipperKickerStatus
+{
+    /* Frankie Robot Fields */
+    // Indicates the time elapsed since chipper was last fired in ms
+    // UINT64_MAX indicates chipper was never fired
+    optional uint64 time_since_chipper_fired = 2;
```
Comment (akhilveeraghanta) [inline src/shared/proto/robot_component_status.proto]: we should adjust these to to be 1,2,3

technically all field numbers until 15 we are safe, but might as well keep it in order

File: src/shared/proto/robot_component_status.proto (hunk 1 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto]: I don't think it's useful so would rather remove it.

File: src/shared/proto/tbots_robot_msg.proto (hunk 3 above)
Comment (MathewMacDougall) [reply src/shared/proto/tbots_robot_msg.proto]: I'm fine with leaving _some_ things required. The reason is that compared to the corporate users of protobuf, we care relatively less about long-term backwards compatibility and we control everything ourselves.

But in the interest of flexibility and recommended style, I have no problems with most/all things being optional. I think we're safe to leave a few things that we are _confident_ would never change, like battery and cap voltage, but if this would make the structure weird then IMO it's not worth it.

File: src/shared/proto/robot_component_status.proto [hunk 6]
```diff
@@ -0,0 +1,132 @@
+syntax = "proto2";
+
+/* Data about the status of the break beam */
+message BreakBeamStatus
+{
+    // Whether or not the ball is interrupting the robot’s laser beam.
+    required bool ball_in_beam = 1;
+
+    // The reading of the robot’s laser sensor.
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Can we add units / a range to this value?

File: src/shared/proto/robot_component_status.proto [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+message DribblerStatus
+{
+    optional int32 dribbler_rpm = 1;
+}
+
+/* Error codes are to be displayed in the visualizer */
+enum ErrorCode
+{
+    /** 2014 ROBOT Messages **/
+    // Electrical messages
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: nit: I don't think we need the `Electrical Messages` heading since a lot more of these errors are electrical. I think splitting by level-triggered and edge-triggered is more important.

nit 2: Most of these error codes apply to "all" robots, so I also don't think we need the `2014 Robot messages` header

File: src/shared/proto/robot_component_status.proto [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    // Indicates the time elapsed since the last primitive packet was received in ms
+    // UINT64_MAX indicates no primitive packet was ever received
+    optional uint64 time_since_last_primitive_received = 3;
+}
+
+/* Battery and capacitor voltages */
+message PowerStatus
+{
+    // The voltage on the robot’s battery, in volts.
+    required double battery_voltage = 1;
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Given the expected ranges of these values I think they can both be `floats` to save space since we don't need `double` precision

File: src/shared/proto/robot_component_status.proto [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    // voltage on the robot’s capacitor, in volts.
+    required double capacitor_voltage = 2;
+}
+
+/* Temperature of robot components */
+message TemperatureStatus
+{
+    // The temperature of the robot’s dribbler motor, in degrees Celsius.
+    required double dribbler_temperature = 1;
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Given the expected ranges of these values I think they can both be `floats` to save space since we don't need `double` precision

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Couple small comments but this is looking good :fire:

File: src/shared/proto/robot_component_status.proto (hunk 1 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto]: JK I have changed my mind, I actually think the `time_since_fired...` values are better than `autokick_fired`. Assuming we use these to help determine when robots have kicked the ball (for things like when an Action is done), the `time_since_fired` values are better because we can report something as fired if we receive any message with the `time_since` value being less than some threshold, indicating the chicker was fired in the last `n` seconds. This makes it easy to tune on the software side.

This is better because if we just had `autokick_fired`, we don't really know how long after the chicker fired that bool will be true. If we miss a few messages or the bool isn't true for very long, we may "miss" the chicker firing.

tl;dr I say keep the `time_since` values and remove `autokick_fired`

File: src/shared/proto/robot_component_status.proto (hunk 2 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto]: I don't think we need this in addition to `fw_build_id`. At best we should make this an error that goes in the error codes.

Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto]: agreed.

File: src/shared/proto/robot_component_status.proto [hunk 10]
```diff
@@ -0,0 +1,132 @@
+syntax = "proto2";
```
Comment (garethellis0) [inline src/shared/proto/robot_component_status.proto]: I think I mentioned this somewhere else, but can't seem to find it now, why `proto2` over `proto3`?

File: src/shared/proto/robot_component_status.proto [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    CRASHED_CORE_DUMPED                  = 27;
+    CRASHED_NO_CORE_DUMP                 = 28;
+
+    /** 2014 DONGLE Messages **/
+    ESTOP_BROKEN_MESSAGE        = 29;
+    RX_FCS_FAIL_MESSAGE         = 30;
+    SECOND_DONGLE_MESSAGE       = 31;
+    TRANSMIT_QUEUE_FULL_MESSAGE = 32;
+    RECEIVE_QUEUE_FULL_MESSAGE  = 33;
+}
```
Comment (garethellis0) [inline src/shared/proto/robot_component_status.proto:78]: Just leaving a quick open comment here to make sure we have someone from elec look  at this before it merges.

File: src/shared/proto/robot_component_status.proto [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    WHEEL_0_HALL_SENSOR_STUCK_HIGH  = 20;
+    WHEEL_1_HALL_SENSOR_STUCK_HIGH  = 21;
+    WHEEL_2_HALL_SENSOR_STUCK_HIGH  = 22;
+    WHEEL_3_HALL_SENSOR_STUCK_HIGH  = 23;
+    DRIBBLER_HALL_SENSOR_STUCK_HIGH = 24;
+    // Edge triggered events
+    ICB_CRC_ERROR                        = 25;
+    RECEIVE_FRAME_CHECK_SEQUENCE_FAILURE = 26;
+    CRASHED_CORE_DUMPED                  = 27;
+    CRASHED_NO_CORE_DUMP                 = 28;
```
Comment (garethellis0) [inline src/shared/proto/robot_component_status.proto]: 25-28 inclusive here are not electrical messages IMO. Perhaps a a separate `SoftwareStatus` proto?

File: src/shared/proto/robot_component_status.proto [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    SECOND_DONGLE_MESSAGE       = 31;
+    TRANSMIT_QUEUE_FULL_MESSAGE = 32;
+    RECEIVE_QUEUE_FULL_MESSAGE  = 33;
+}
+
+/* Firmware build information  */
+message FirmwareStatus
+{
+    // The microcontroller firmware build ID.
+    required uint32 fw_build_id = 1;
```
Comment (garethellis0) [inline src/shared/proto/robot_component_status.proto]: IIRC this is just set in code, but using `firmware_binary_hash` (and actually implementing that) or something similar would make much more sense to me.

Certainly out-of-scope here, but just noting it while I'm looking at it.

File: src/shared/proto/robot_component_status.proto [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+/* Data from the motor encoders */
+message MotorEncoderData
+{
+    optional int32 motor_fl_speed = 1;
+    optional int32 motor_fr_speed = 2;
+    optional int32 motor_bl_speed = 3;
+    optional int32 motor_br_speed = 4;
+}
```
Comment (garethellis0) [inline src/shared/proto/robot_component_status.proto:104]: Units on all of these would be good.

File: src/shared/proto/tbots_robot_msg.proto (hunk 3 above)
Comment (garethellis0) [reply src/shared/proto/tbots_robot_msg.proto]: If something does not need to be optional it shouldn't be, as optional things incur more syntactic and semantic overhead for the developer working with them.

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/shared/proto/robot_component_status.proto (hunk 1 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 2 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 5 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 10 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: um no good reason, changed it

File: src/shared/proto/robot_component_status.proto (hunk 9 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 8 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 7 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 6 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/tbots_robot_msg.proto (hunk 3 above)
Comment (jonathanlew) [reply src/shared/proto/tbots_robot_msg.proto]: turns out proto3 doesn't support required so we're stuck with optional

File: src/shared/proto/robot_component_status.proto (hunk 14 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto:104]: done

Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto:104]: Could we have them on the name? ie. `motor_br_speed_m_per_s`?

Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto:104]: yeah, also it's a placeholder for the sort of encoder data we actually want

**Review by akhilveeraghanta — COMMENTED** [review]
LGTM this is looking really clean, switching over to proto3 was a good idea.

File: src/shared/proto/robot_component_status.proto [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    // The reading of the robot’s laser sensor, normalized to [0,1]
+    float break_beam_reading = 2;
+}
+
+/* Data about the status of the chipper and kicker */
+message ChipperKickerStatus
+{
+    // Indicates the time elapsed since chipper was last fired in ms
+    // UINT64_MAX indicates chipper was never fired
+    uint64 time_since_chipper_fired = 1;
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto:18]: I know using the unsigned version of whatever the largest int type available is pretty "standard" for time-based stuff, but a `uint32` would still represent nearly 50 days of time. This might be nitpicking at this point, but with the possibility of using these protos over radio (I know that's undecided atm) we may want to consider using `uint32` for saving space.

**Edit:** Although I suppose since it's encoded as a varint it doesn't matter

File: src/shared/proto/robot_component_status.proto [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+    // Indicates the time elapsed since kicker was last fired in ms
+    // UINT64_MAX indicates kicker was never fired
+    uint64 time_since_kicker_fired = 2;
+}
+
+/* Data about the status of the dribbler */
+message DribblerStatus
+{
+    int32 dribbler_rpm = 1;
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: Should this be unsigned? Assigning a sign to a direction here is fine, just pointing this out to be explicit. If we do want to assign a sign to a direction we should document it.

File: src/shared/proto/robot_component_status.proto [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+/* Error codes are to be displayed in the visualizer */
+enum ErrorCode
+{
+    NO_ERROR                = 0;  // default message
+    ROBOT_DEAD_MESSAGE      = 1;
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: `ROBOT_DEAD_MESSAGE` is a radio dongle-specific message. Perhaps it's worth putting any radio dongle errors in their own heading / group since they may be removed when radio support is dropped? (Just to make it easy to identify). This would go with things like
* `SECOND_DONGLE_MESSAGE`
* `ESTOP Broken`
* `TRANSMIT_QUEUE_FULL_MESSAGE`
etc.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just a few nits but otherwise this is looking great!

File: src/shared/proto/robot_component_status.proto (hunk 16 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: added more comments

File: src/shared/proto/robot_component_status.proto (hunk 17 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: I don't think the complexity of another message isn't worth it. We can probably remove those enum values if we will never run old robots again. I added more comments.

File: src/shared/proto/robot_component_status.proto (hunk 16 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto]: Can we specify "along the shaft" or use the working we used for other motors in the new `app` interfaces? I can't seem to find the comments / exact wording so summoning @garethellis0

File: src/shared/proto/robot_component_status.proto [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    // NOTE: positive means clockwise and negative means counterclockwise
+    int32 dribbler_rpm = 1;
+}
+
+/* Error codes are to be displayed in the visualizer */
+enum ErrorCode
+{
+    NO_ERROR                = 0;  // default message
+    LOW_CAP_MESSAGE         = 1;
+    LOW_BATTERY_MESSAGE     = 2;
```
Comment (MathewMacDougall) [inline src/shared/proto/robot_component_status.proto]: nit: these 3 messages are level-triggered so can be in that section

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/shared/proto/robot_component_status.proto (hunk 17 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto]: Yea sorry I didn't mean another message, just grouping in a comment. What you did looks good :+1:

File: src/shared/proto/robot_component_status.proto (hunk 18 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: done

File: src/shared/proto/robot_component_status.proto (hunk 16 above)
Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto]: I think this is what you're thinking of, it's from some IO code that we were waiting on hardware to test (so it's not merged yet):
```
/**
 * Set the rotation direction for the given motor driver
 *
 * Note that the rotation is from the perspective of rear of the motor, looking down
 * the shaft starting from the motor body
 *
 * @param motor_driver
 * @param direction
 */
```

File: src/shared/proto/robot_component_status.proto (hunk 14 above)
Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto:104]: Yeah, totally fine if we need to change later, just don't want to be generic in case people start using it......

File: src/shared/proto/robot_component_status.proto (hunk 16 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto]: Yea this is what I was thinking of

File: src/shared/proto/robot_component_status.proto (hunk 15 above)
Comment (akhilveeraghanta) [reply src/shared/proto/robot_component_status.proto:18]: Do we have a use-case for these values?

File: src/shared/proto/robot_component_status.proto [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+{
+    // The microcontroller firmware build ID.
+    uint32 fw_build_id = 1;
+}
+
+/* Data from the motor encoders */
+message MotorEncoderData
+{
+    // NOTE: positive means clockwise and negative means counterclockwise
+    int32 motor_fl_rpm = 1;
```
Comment (akhilveeraghanta) [inline src/shared/proto/robot_component_status.proto]: uber nit: can we expand the abbreviations? 

These values are in `MotorEncoderData` so its obvious that they are for `motor`s so we can probably remove the `motor` and just expand fl,br,bl,br to front left, back right, etc...

`motor_fr_rpm` -> `motor_front_right_rpm` or even `front_right_rpm`

File: src/shared/proto/robot_component_status.proto (hunk 15 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto:18]: From a previous comment. tl;dr, yea as a better `autokick_fired` metric, which is useful for software detecting kicks (for done conditions).

> JK I have changed my mind, I actually think the time_since_fired... values are better than autokick_fired. Assuming we use these to help determine when robots have kicked the ball (for things like when an Action is done), the time_since_fired values are better because we can report something as fired if we receive any message with the time_since value being less than some threshold, indicating the chicker was fired in the last n seconds. This makes it easy to tune on the software side.

> This is better because if we just had autokick_fired, we don't really know how long after the chicker fired that bool will be true. If we miss a few messages or the bool isn't true for very long, we may "miss" the chicker firing.

> tl;dr I say keep the time_since values and remove autokick_fired

Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto:18]: I've removed autokick_fired

File: src/shared/proto/robot_component_status.proto (hunk 16 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: Added 
* Note that the rotation is from the perspective of rear of the motor, looking down
 * the shaft starting from the motor body
as a comment

File: src/shared/proto/robot_component_status.proto (hunk 19 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: changed to front_right_rpm

File: src/shared/proto/robot_component_status.proto [hunk 20]
```diff
@@ -0,0 +1,131 @@
+syntax = "proto3";
+
+/* Data about the status of the break beam */
+message BreakBeamStatus
+{
+    // Whether or not the ball is interrupting the robot’s laser beam.
+    bool ball_in_beam = 1;
+
+    // The reading of the robot’s laser sensor, normalized to [0,1]
+    float break_beam_reading = 2;
```
Comment (EvanMorcom) [inline src/shared/proto/robot_component_status.proto:10]: Similar to below, why are we sending both the 'ball_in_break_beam' and 'break_beam_reading'? Could you not get the 'ball in break beam' data from the reading?

File: src/shared/proto/robot_component_status.proto [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+enum ErrorCode
+{
+    // Default message required for proto enums
+    NO_ERROR = 0;
+
+    // Level triggered events
+    LOW_CAP_MESSAGE                 = 1;
+    LOW_BATTERY_MESSAGE             = 2;
+    HIGH_BOARD_TEMP_MESSAGE         = 3;
+    CHARGE_TIMEOUT                  = 4;
```
Comment (EvanMorcom) [inline src/shared/proto/robot_component_status.proto:44]: Any reason why the first 3 messages contain '_MESSAGE' but the rest don't?

File: src/shared/proto/robot_component_status.proto (hunk 11 above)
Comment (EvanMorcom) [reply src/shared/proto/robot_component_status.proto:78]: @hannahvsawiuk

File: src/shared/proto/robot_component_status.proto [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    WHEEL_1_HALL_SENSOR_STUCK_LOW   = 15;
+    WHEEL_2_HALL_SENSOR_STUCK_LOW   = 16;
+    WHEEL_3_HALL_SENSOR_STUCK_LOW   = 17;
+    DRIBBLER_HALL_SENSOR_STUCK_LOW  = 18;
+    WHEEL_0_HALL_SENSOR_STUCK_HIGH  = 19;
+    WHEEL_1_HALL_SENSOR_STUCK_HIGH  = 20;
+    WHEEL_2_HALL_SENSOR_STUCK_HIGH  = 21;
+    WHEEL_3_HALL_SENSOR_STUCK_HIGH  = 22;
+    DRIBBLER_HALL_SENSOR_STUCK_HIGH = 23;
+    // Edge triggered events
```
Comment (EvanMorcom) [inline src/shared/proto/robot_component_status.proto]: Nit:

You don't have a space after this section, but have one before the radio messages

File: src/shared/proto/robot_component_status.proto [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+    CRASHED_CORE_DUMPED                  = 26;
+    CRASHED_NO_CORE_DUMP                 = 27;
+
+    // Radio Dongle Errors
+    ROBOT_DEAD_MESSAGE          = 28;
+    ESTOP_BROKEN_MESSAGE        = 29;
+    RX_FCS_FAIL_MESSAGE         = 30;
+    SECOND_DONGLE_MESSAGE       = 31;
+    TRANSMIT_QUEUE_FULL_MESSAGE = 32;
+    RECEIVE_QUEUE_FULL_MESSAGE  = 33;
```
Comment (EvanMorcom) [inline src/shared/proto/robot_component_status.proto]: Similar question about the appending of '_MESSAGE' here but not in other messages

File: src/shared/proto/robot_component_status.proto (hunk 20 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto:10]: It might be good for debugging to report break_beam_reading, but for actual use, we only need ball_in_break_beam

File: src/shared/proto/robot_component_status.proto (hunk 23 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: ok removed _MESSAGE

File: src/shared/proto/robot_component_status.proto [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+/* Data from the motor encoders
+ * Note that the rotation is from the perspective of rear of the motor, looking down
+ * the shaft starting from the motor body
+ */
+message MotorEncoderData
+{
+    int32 front_left_rpm  = 1;
+    int32 front_right_rpm = 2;
+    int32 back_left_rpm   = 3;
+    int32 back_right_rpm  = 4;
```
Comment (EvanMorcom) [inline src/shared/proto/robot_component_status.proto]: Encoders measure position data primarily, I think It would be more clear to rename this to MotorEncoderVelocity or something similar. 

It would also be nice to have access to the encoder position data as well. Maybe bundle them together in another message and make the units rad/s and rad.

@garethellis0 thoughts?

**Review by EvanMorcom — DISMISSED** [review]
:fire:

LGTM pending a few minor comments!

File: src/shared/proto/robot_component_status.proto (hunk 24 above)
Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto]: Yeah, something like:
```
message DriveUnit {
float wheel_rpm = 1;
float wheel_position_rad = 2;
}

message DriveUnits {
DriveUnit front_left = 1;
DriveUnit front_right = 2;
DriveUnit back_left = 3;
DriveUnit back_right = 4;
}
```
??

**Review by hannahvsawiuk — DISMISSED** [review]

File: src/shared/proto/robot_component_status.proto (hunk 24 above)
Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto]: OK I've added that

**Review by akhilveeraghanta — APPROVED** [review]
LGTM, just realized this will not build w/ the existing nanopb macro because the msg is broken into two different files, but that's out of the scope for this PR, and will be solved by #1133

File: src/shared/proto/robot_component_status.proto (hunk 24 above)
Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto]: Seems OK to me now, I might say that it should be called be `DriveUnits` instead of `DriveEncoderData` because technically the wheel rpm and position takes into account the gear ratios of the drive unit as well, so it's not just encoder data, but I'm fine either way.

**Review by garethellis0 — APPROVED** [review]

File: src/shared/proto/robot_component_status.proto [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+    uint32 fw_build_id = 1;
+}
+
+/* Encoder data from a single motor
+ * Note that the rotation is from the perspective of rear of the motor, looking down
+ * the shaft starting from the motor body
+ */
+message DriveUnit
+{
+    float wheel_rpm          = 1;
```
Comment (EvanMorcom) [inline src/shared/proto/robot_component_status.proto:93]: Very nit:

Can we get wheel_ang_vel (in rad/s) instead of RPM? It's just more usable and I see RPM being converted anyways.

**Review by EvanMorcom — APPROVED** [review]
LGTM :fire: 🚂

File: src/shared/proto/robot_component_status.proto (hunk 15 above)
Comment (MathewMacDougall) [reply src/shared/proto/robot_component_status.proto:18]: @akhilveeraghanta @jonathanlew @garethellis0 thoughts on `uint64` vs `uint32`? Leave it until we for sure decide to send proto over radio?

Comment (jonathanlew) [reply src/shared/proto/robot_component_status.proto:18]: > Leave it until we for sure decide to send proto over radio?

That sounds good. I think there are some minor issues left with TbotsRobotMsg that can be dealt with much more easily once we implement the related pieces in firmware and sensor fusion.

Comment (garethellis0) [reply src/shared/proto/robot_component_status.proto:18]: I think that we should leave it as `uint64` given that we're using varint's (as you say) and thus there shouldn't be any size difference.

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:
